crypto/rand/rand_unix.c

```diff
@@ -441,16 +441,21 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
     unsigned char *buffer;
 
 #   ifdef OPENSSL_RAND_SEED_GETRANDOM
-    bytes_needed = rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
-    buffer = rand_pool_add_begin(pool, bytes_needed);
-    if (buffer != NULL) {
-        size_t bytes = 0;
-
-        if (syscall_random(buffer, bytes_needed) == (int)bytes_needed)
-            bytes = bytes_needed;
+    {
+        int bytes, attempts = 3;
 
-        rand_pool_add_end(pool, bytes, 8 * bytes);
-        entropy_available = rand_pool_entropy_available(pool);
+        bytes_needed = rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
+        while(bytes_needed != 0 && attempts-- > 0) {
+            buffer = rand_pool_add_begin(pool, bytes_needed);
+            bytes = syscall_random(buffer, bytes_needed);
+            if (bytes > 0) {
+                rand_pool_add_end(pool, (size_t)bytes, 8 * (size_t)bytes);
```

Comment:
Here is interesting question. Is explicit cast actually required? And if not, are we better off without it? Cast is not actually required because of explicit bytes > 0 check, and implicit conversion would work predictably. But are we better off with it? I'm **not** saying that I have an answer, but I can't help noting that with unsigned multiplication being specified as modulo word-size operation, and signed can turning undefined if multiplicands are large enough, explicit cast renders outcome of multiplication by 8 unconditionally valid. And without it we would give undefined behaviour sanitizer to catch it for us... Does it mean that a check for result of multiplication to be guaranteed appropriate? Again, not that I know the definitive answer, but *formally speaking* yes...

Reply:
> Here is interesting question. Is explicit cast actually required?

I really love your interesting questions. Very often I learn a lot when following the discussions they trigger :-). I must admit that it was a pure reflex that I placed the cast in anticipation of a compiler warning. You were right, there is no warning about sign conversion when I remove the casts. But the reason you gave seems to be incorrect, there is also no warning even if I replace `if (bytes > 0)` by  `if (bytes < 0)`. The true cause seems to be that `-Wsign-conversion` is disabled even when configuring  `--strict-warnings`. Here's my configuration:

```
/usr/bin/perl ./Configure debug-linux-x86_64 threads shared enable-egd enable-crypto-mdebug -g --strict-warnings --prefix=/home/msp/install/openssl
```

I tried to add `-Wsign-conversion` explicitely and quickly gave up, because there were tons of warnings.

Reply:
But putting compilers and sanitizers aside: Yes, I agree with you that because of the if-clause it is safe to omit the cast.

Reply:
On non-direct, vaguely related note, one can only wonder how did they reason declaring call as one taking size_t and returning int. Of course that large lengths are not something anybody would ever ask for, but declaration asks for question what happens if you ask for 4GB? Would it return 2GB-1? Again, just a side note...

Reply:
@dot-asm please have a look at 23d7c0985115f412b0cc117fafd928f7d575bb7f where I addressed your comment about the return value of syscall_random().

Reply:
Note that I didn't actually suggest to change the declaration, only wondered what did **they** think when declaring underlying call like that. I mean syscall_random is our function, but it calls system-specific one that accepts size_t and returns int, and remark was about the latter. But now when I look at it in morning light, I realize that it was misdirected, because underlying call doesn't return amount of bytes, but error or success. Duh! This however makes question applicable to syscall_random...